pkg-manager/resolve-dependencies/src/resolveDependencies.ts

```diff
@@ -1185,6 +1185,17 @@ async function resolveDependency (
     }
   }
   try {
+    // For dependencies of importers, the catalog protocol should be substituted
+    // by this point. If a dependency pref using the catalog protocol is
+    // encountered this deep into resolution, it's likely because it's an out of
+    // repo dependency that was published incorrectly. For example, it may be
+    // been mistakenly published with 'npm publish' instead of 'pnpm publish'.
+    if (wantedDependency.pref.startsWith('catalog:')) {
```

Comment:
This is a convenient place to throw an error since the `catch` block sets up `pkgsStack`, but let me know if we think this error should be thrown somewhere else.

Reply:
maybe do the check only when there's an error.

Reply:
I'm wondering if we should reuse existing `ERR_PNPM_SPEC_NOT_SUPPORTED_BY_ANY_RESOLVER` error instead and detect if the package that was unsupported was the `catalog:` protocol. I realized this might be simpler after looking at this code more today.

I just pushed up this approach as https://github.com/pnpm/pnpm/pull/8254/commits/438d0b900c5f2c0ae24a391d35385711ab35409b, which removes the throw inside `resolveDependencies.ts` entirely.

What do you think?

<img width="842" alt="Screenshot 2024-06-29 at 12 37 22 AM" src="https://github.com/pnpm/pnpm/assets/906558/8c6308a6-69de-4d97-91a5-22ef87b88f7d">